File: step_11_heaps/11.2_medium/IB120_magician_and_chocolates/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE
// Time: O(K * N log N)  |  Space: O(N)
// Simulate K rounds: sort bags each round, pick max, halve, reinsert
// ============================================================
class BruteForce {
    public static long solve(int[] bags, int k) {
        int n = bags.length;
        int[] b = bags.clone();
        long total = 0;
        for (int t = 0; t < k; t++) {
            Arrays.sort(b);
            int maxVal = b[n - 1];
            total += maxVal;
            b[n - 1] = maxVal / 2;
        }
        return total;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - Max-Heap
// Time: O(K log N)  |  Space: O(N)
// Use a max-heap: each turn extract max, add to profit, push max/2
// ============================================================
class Optimal {
    public static long solve(int[] bags, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int b : bags) pq.offer(b);
        long total = 0;
        for (int t = 0; t < k; t++) {
            int maxVal = pq.poll();
            total += maxVal;
            pq.offer(maxVal / 2);
        }
        return total;
    }
}

// ============================================================
// APPROACH 3: BEST - Max-Heap with early exit
// Time: O(K log N)  |  Space: O(N)
// Same as Optimal but stop early if max bag is 0
// ============================================================
class Best {
    public static long solve(int[] bags, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int b : bags) pq.offer(b);
        long total = 0;
        for (int t = 0; t < k; t++) {
            int maxVal = pq.peek();
            if (maxVal == 0) break;  // all bags empty
            pq.poll();
            total += maxVal;
            pq.offer(maxVal / 2);
        }
        return total;
    }
}

public class Solution {
    public static void main(String[] args) {
        // IB Example: A=[6,5], k=3 -> 6+5+3=14 (6->3, 5->2; next max=3->1; next=2)
        // Actually: round1: max=6, total=6, push 3. Heap: [5,3]
        //           round2: max=5, total=11, push 2. Heap: [3,2]
        //           round3: max=3, total=14, push 1. Answer=14
        int[] bags1 = {6, 5};
        System.out.println("Test 1 (bags=[6,5], k=3):");
        System.out.println("  BruteForce = " + BruteForce.solve(bags1, 3));  // 14
        System.out.println("  Optimal    = " + Optimal.solve(bags1, 3));     // 14
        System.out.println("  Best       = " + Best.solve(bags1, 3));        // 14

        int[] bags2 = {1, 2, 3};
        System.out.println("Test 2 (bags=[1,2,3], k=4):");
        System.out.println("  BruteForce = " + BruteForce.solve(bags2, 4));
        System.out.println("  Optimal    = " + Optimal.solve(bags2, 4));
        System.out.println("  Best       = " + Best.solve(bags2, 4));
        // round1: max=3, total=3, push 1. [2,1,1]
        // round2: max=2, total=5, push 1. [1,1,1]
        // round3: max=1, total=6, push 0. [1,1,0]
        // round4: max=1, total=7, push 0. => 7
    }
}

File: step_11_heaps/11.2_medium/IB120_magician_and_chocolates/solution.py
"""Problem: Magician and Chocolates
Difficulty: EASY | XP: 10
Source: InterviewBit

K time units, N bags of chocolates. Each time unit: take the bag with most chocolates,
eat them all, put back floor(eaten/2). Maximize total chocolates eaten in K turns.
"""
import heapq
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(K * N log N)  |  Space: O(N)
# Sort each round and pick max manually
# ============================================================
def brute_force(bags: List[int], k: int) -> int:
    b = bags[:]
    total = 0
    for _ in range(k):
        b.sort()
        max_val = b[-1]
        total += max_val
        b[-1] = max_val // 2
    return total % (10**9 + 7)


# ============================================================
# APPROACH 2: OPTIMAL - Max-Heap
# Time: O(K log N)  |  Space: O(N)
# Use max-heap (negate for Python min-heap)
# ============================================================
def optimal(bags: List[int], k: int) -> int:
    MOD = 10**9 + 7
    heap = [-x for x in bags]
    heapq.heapify(heap)
    total = 0
    for _ in range(k):
        max_val = -heapq.heappop(heap)
        total += max_val
        heapq.heappush(heap, -(max_val // 2))
    return total % MOD


# ============================================================
# APPROACH 3: BEST - Max-Heap with early exit when all zero
# Time: O(K log N)  |  Space: O(N)
# Stop early once the max bag is 0 (no more chocolates)
# ============================================================
def best(bags: List[int], k: int) -> int:
    MOD = 10**9 + 7
    heap = [-x for x in bags]
    heapq.heapify(heap)
    total = 0
    for _ in range(k):
        max_val = -heap[0]
        if max_val == 0:
            break
        heapq.heapreplace(heap, -(max_val // 2))
        total += max_val
    return total % MOD


if __name__ == "__main__":
    tests = [
        ([6, 5], 3, 14),       # 6+5+3=14
        ([1, 2, 3], 4, 7),     # 3+2+1+1=7
        ([10], 1, 10),
        ([0, 0], 5, 0),        # all zero bags
    ]
    for bags, k, expected in tests:
        bf = brute_force(bags, k)
        opt = optimal(bags, k)
        be = best(bags, k)
        status = "OK" if bf == opt == be == expected else "FAIL"
        print(f"[{status}] bags={bags} k={k} -> Brute={bf}, Optimal={opt}, Best={be} (expected={expected})")
